src/ordered_robin_dict.jl

```diff
@@ -0,0 +1,476 @@
+        return h
+    end
+    OrderedRobinDict{K,V}(p::Pair) where {K,V} = setindex!(OrderedRobinDict{K,V}(), p.second, p.first)
+    function OrderedRobinDict{K,V}(ps::Pair...) where V where K
+        h = OrderedRobinDict{K,V}()
+        sizehint!(h, length(ps))
+        for p in ps
+            h[p.first] = p.second
+        end
+        return h
+    end
+end
+
+OrderedRobinDict() = OrderedRobinDict{Any,Any}()
+OrderedRobinDict(kv::Tuple{}) = OrderedRobinDict()
+copy(d::OrderedRobinDict) = OrderedRobinDict(d)
+empty(d::OrderedRobinDict, ::Type{K}, ::Type{V}) where {K, V} = OrderedRobinDict{K, V}()
+
+OrderedRobinDict(ps::Pair{K,V}...) where {K,V} = OrderedRobinDict{K,V}(ps)
+OrderedRobinDict(ps::Pair...)                  = OrderedRobinDict(ps)
+
+OrderedRobinDict(d::AbstractDict{K, V}) where {K, V} = OrderedRobinDict{K, V}(d)
+
+function OrderedRobinDict(kv)
+    try
+        return dict_with_eltype((K, V) -> OrderedRobinDict{K, V}, kv, eltype(kv))
+    catch e
+        if !isiterable(typeof(kv)) || !all(x -> isa(x, Union{Tuple,Pair}), kv)
+            !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("OrderedRobinDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
+end
+
+empty(d::OrderedRobinDict{K,V}) where {K,V} = OrderedRobinDict{K,V}()
+
+length(d::OrderedRobinDict) = d.count
+isempty(d::OrderedRobinDict) = (length(d) == 0)
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = OrderedRobinDict("a" => 1, "b" => 2)
+OrderedRobinDict{String,Int64} with 2 entries:
+  "a" => 1
+  "b" => 2
+
+julia> empty!(A);
+
+julia> A
+OrderedRobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::OrderedRobinDict{K,V}) where {K, V}
+    empty!(h.dict)
+    empty!(h.keys)
+    empty!(h.vals)
+    h.count = 0
+    return h
+end
+
+function _setindex!(h::OrderedRobinDict, v, key)
+    hk, hv = h.keys, h.vals
+    push!(hk, key)
+    push!(hv, v)
+    nk = length(hk)
+    @inbounds h.dict[key] = Int32(nk)
+    h.count += 1
+end
+
+function setindex!(h::OrderedRobinDict{K, V}, v0, key0) where {K,V}
+    key = convert(K,key0)
+
+    v = convert(V,  v0)
+
+    index = get(h.dict, key, -2)
+
+    if index < 0
+        _setindex!(h, v0, key0)
+    else
+        @assert haskey(h, key0)
+        @inbounds orig_v = h.vals[index]
+        (orig_v != v0) && (@inbounds h.vals[index] = v0)
+    end
+
+    check_for_rehash(h) && rehash!(h)
+
+    return h
+end
+
+# rehash when there are ALLOWABLE_USELESS_GROWTH %
+# tombstones, or non-mirrored entries in the dictionary
+function check_for_rehash(h::OrderedRobinDict)
+    keysl = length(h.keys)
+    dictl = length(h)
+    return (keysl > (1 + ALLOWABLE_USELESS_GROWTH)*dictl)
+end
+
+function rehash!(h::OrderedRobinDict{K, V}) where {K, V}
+    keys = h.keys
+    vals = h.vals
+    newk = Vector{K}()
+    newv = Vector{V}()
+    hk, hv = newk, newv
```

Comment:
why do we need `hk` and `hv` ?
can't we just use `newk` and `newv` since they are aliases?